src/dragdrop/ranking-select-to-rank.ts

```diff
@@ -0,0 +1,163 @@
+import { ItemValue } from "../itemvalue";
+import { DragDropRankingChoices } from "./ranking-choices";
+import { QuestionRankingModel } from "../question_ranking";
+
+export class DragDropRankingSelectToRank extends DragDropRankingChoices {
+  protected findDropTargetNodeByDragOverNode(
+    dragOverNode: HTMLElement
+  ): HTMLElement {
+    if (dragOverNode.dataset.ranking === "from-container" || dragOverNode.dataset.ranking === "to-container") {
+      return dragOverNode;
+    }
+
+    if (this.parentElement.isEmpty()) {
+      let toContainer: HTMLElement = dragOverNode.closest("[data-ranking='to-container']");
+      let fromContainer: HTMLElement = dragOverNode.closest("[data-ranking='from-container']");
+
+      if (!!toContainer) {
+        return toContainer;
+      } else if (!!fromContainer) {
+        return fromContainer;
+      } else {
+        return null;
+      }
+    }
+
+    return super.findDropTargetNodeByDragOverNode(dragOverNode);
+  }
+
+  protected getDropTargetByDataAttributeValue(dataAttributeValue: string): ItemValue {
+    return this.parentElement.rankingChoices[dataAttributeValue] || this.parentElement.unRankingChoices[dataAttributeValue];
+  }
+
+  protected getDropTargetByNode(
+    dropTargetNode: HTMLElement,
+    event: PointerEvent
+  ): any {
+    if (dropTargetNode.dataset.ranking === "to-container") {
+      return "to-container";
+    }
+
+    if (dropTargetNode.dataset.ranking === "from-container" || dropTargetNode.closest("[data-ranking='from-container']")) {
+      return "from-container";
+    }
+
+    return super.getDropTargetByNode(dropTargetNode, event);
+  }
+
+  protected isDropTargetValid(
+    dropTarget: ItemValue | string,
+    dropTargetNode?: HTMLElement
+  ): boolean {
+    if (dropTarget === "to-container" || dropTarget === "from-container") {
+      return true;
+    } else {
+      return super.isDropTargetValid(<ItemValue>dropTarget, dropTargetNode);
+    }
+  }
+
+  protected afterDragOver(dropTargetNode: HTMLElement): void {
+    const questionModel: any = this.parentElement;
+    const rankingChoices = questionModel.rankingChoices;
+    const unRankingChoices = questionModel.unRankingChoices;
+
+    let fromIndex;
+    let toIndex;
+
+    if (this.isDraggedElementUnranked && this.isDropTargetRanked) {
+      fromIndex = unRankingChoices.indexOf(this.draggedElement);
+      if (rankingChoices.length === 0) {
+        toIndex = 0;
+      } else {
+        toIndex = rankingChoices.indexOf(this.dropTarget);
+      }
+      this.selectToRankEnabled(questionModel, fromIndex, toIndex);
+      this.doUIEffects(dropTargetNode, fromIndex, toIndex);
+      return;
+    }
+
+    if (this.isDraggedElementRanked && this.isDropTargetRanked) {
```

Comment:
Code duplication, create a function that will pass choices (ranking or unranking) and function one of the function: selectToRankEnabled or reorderRankedItem or unselectFromRank.